File: apps/desktop/src/lib/timeline-purge.test.ts
// Tests for book 16 §6.2 (clearing) and book 15 G-21 (a single-row delete must not
// delete bytes).
//
// The load-bearing ones are the "same delete path" pair: a single-row delete and a clear
// must have the SAME side effects on the same row, because before C2 they did not —
// eviction dropped the picture file and the user's delete left it on disk forever.

import { beforeEach, describe, expect, it, vi } from 'vitest';
import { MAX_ROWS } from './timeline-retention';
import {
  CLEAR_WINDOWS,
  NO_CUTOFFS,
  advanceCutoffs,
  combinedCutoff,
  horizonOf,
  planClear,
  purge,
  readCutoffs,
  type Cutoffs,
  type PurgeSink,
} from './timeline-purge';
import { TimelineStore } from './timeline-store';
import type { ReportingKvStore, TimelineRow, TimelineTransport } from './types';

const NOW = Date.parse('2026-08-02T12:00:00.000Z');

function row(over: Partial<TimelineRow> & { id: string }): TimelineRow {
  return {
    mode: 'realtime',
    status: 'injected',
    edited: false,
    source_text: null,
    output_text: 'hello',
    created_at: '2026-08-01T00:00:00.000Z',
    updated_at: '2026-08-01T00:00:00.000Z',
    entry_type: 'transcript',
    thumb_b64: null,
    full_image: false,
    target: null,
    mobile_id: null,
    mobile_label: null,
    channel: 'lan',
    ...over,
  } as TimelineRow;
}

function sinkOf(rows: TimelineRow[]): { sink: PurgeSink; dropped: string[][] } {
  const dropped: string[][] = [];
  const map = new Map(rows.map((r) => [`${r.channel}:${r.id}`, r]));
  return {
    dropped,
    sink: {
      rows: map,
      imageIds: new Set(rows.map((r) => r.id)),
      dropPictures: (ids) => dropped.push([...ids]),
      keyOf: (r) => `${r.channel}:${r.id}`,
    },
  };
}

describe('planClear — 二选 + 时间档 (owner 2026-08-01 §4-4)', () => {
  const rows = [
    row({ id: 'old-text', created_at: '2026-01-01T00:00:00.000Z' }),
    row({ id: 'new-text', created_at: '2026-08-02T00:00:00.000Z' }),
    row({ id: 'old-pic', created_at: '2026-01-01T00:00:00.000Z', entry_type: 'image', full_image: true }),
    row({ id: 'new-pic', created_at: '2026-08-02T00:00:00.000Z', entry_type: 'image', full_image: true }),
  ];

  it('「清图片 · 一月以前」 takes only old pictures — the other three are untouched', () => {
    const doomed = planClear(rows, 'images', horizonOf('month', NOW));
    expect(doomed.map((r) => r.id)).toEqual(['old-pic']);
  });

  it('「清文字 · 一月以前」 takes only old transcripts — the mirror of the above', () => {
    const doomed = planClear(rows, 'text', horizonOf('month', NOW));
    expect(doomed.map((r) => r.id)).toEqual(['old-text']);
  });

  it('「全部」 has no lower bound and takes everything of that kind, however recent', () => {
    expect(planClear(rows, 'both', horizonOf('all', NOW))).toHaveLength(4);
    expect(planClear(rows, 'images', horizonOf('all', NOW)).map((r) => r.id)).toEqual(['old-pic', 'new-pic']);
  });

  it('a row with no usable created_at is NOT swept up by a time-bounded clear', () => {
    // '' sorts below every real instant, so a naive `<` would delete a row of
    // genuinely unknown age under "older than a year". It only goes under "all", which
    // makes no claim about age.
    const junk = [row({ id: 'junk', created_at: '' })];
    expect(planClear(junk, 'both', horizonOf('year', NOW))).toHaveLength(0);
    expect(planClear(junk, 'both', horizonOf('all', NOW))).toHaveLength(1);
  });

  it('every bucket except 「全部」 yields a horizon, and they get older in order', () => {
    const marks = CLEAR_WINDOWS.filter((w) => w !== 'all').map((w) => horizonOf(w, NOW));
    expect(marks.every((m) => m !== null)).toBe(true);
    expect([...marks].sort().reverse()).toEqual(marks); // week newest … year oldest
    expect(horizonOf('all', NOW)).toBeNull();
  });
});

describe('purge — 删一行 = 删掉这一行的全部字节 (16 册 §6.2-2)', () => {
  it('removes the row, its picture FILE and its image chip, in one call', () => {
    const pic = row({ id: 'p1', entry_type: 'image', full_image: true });
    const { sink, dropped } = sinkOf([pic, row({ id: 't1' })]);
    const out = purge(sink, [pic], null, NO_CUTOFFS);
    expect(out.removed).toBe(1);
    expect(out.picturesDropped).toBe(1);
    expect(sink.rows.has('lan:p1')).toBe(false);
    expect(dropped).toEqual([['p1']]); // ← the half G-21 was missing
    expect(sink.imageIds.has('p1')).toBe(false);
    expect(sink.rows.has('lan:t1')).toBe(true); // positive control: the other row survived
  });

  it('a picture row that never kept its bytes asks the filesystem for nothing', () => {
    // §9b-6: imageCount ≠ imageFileCount is real history, and a deletion must not
    // claim to free bytes that are not there.
    const pic = row({ id: 'p2', entry_type: 'image', full_image: false });
    const { sink, dropped } = sinkOf([pic]);
    expect(purge(sink, [pic], null, NO_CUTOFFS).picturesDropped).toBe(0);
    expect(dropped).toEqual([]);
  });

  it('a row named twice, or already gone, counts once and drops once', () => {
    const pic = row({ id: 'p3', entry_type: 'image', full_image: true });
    const { sink, dropped } = sinkOf([pic]);
    const out = purge(sink, [pic, pic], null, NO_CUTOFFS);
    expect(out.removed).toBe(1);
    expect(dropped).toEqual([['p3']]);
  });
});

describe('cutoffs — 清掉 ≠ 从未有过, without over-claiming (16 册 §6.2-3)', () => {
  const t = row({ id: 't', created_at: '2026-03-01T00:00:00.000Z' });
  const p = row({ id: 'p', created_at: '2026-04-01T00:00:00.000Z', entry_type: 'image' });

  it('a single-row delete moves NO mark —「我删了这条」≠「早于它的都没了」', () => {
    expect(advanceCutoffs(NO_CUTOFFS, [t], null)).toEqual(NO_CUTOFFS);
  });

  it('clearing only pictures does not let the page claim transcripts are gone too', () => {
    const after = advanceCutoffs(NO_CUTOFFS, [p], 'images');
    expect(after.images).toBe(p.created_at);
    expect(after.text).toBeNull();
    // 🔴 THE ASSERTION THIS WHOLE SHAPE EXISTS FOR: one scalar cutoff would now read
    // "everything before April has been cleared" while every transcript from March is
    // still on screen.
    expect(combinedCutoff(after)).toBeNull();
  });

  it('a kind-blind deletion earns BOTH marks even when it only caught transcripts', () => {
    // Capacity eviction drops oldest-first regardless of kind, so everything older
    // than its newest casualty really is gone — of every kind.
    const after = advanceCutoffs(NO_CUTOFFS, [t], 'both');
    expect(after).toEqual({ text: t.created_at, images: t.created_at });
    expect(combinedCutoff(after)).toBe(t.created_at);
  });

  it('marks only ever move forward', () => {
    const start: Cutoffs = { text: '2026-05-01T00:00:00.000Z', images: '2026-05-01T00:00:00.000Z' };
    expect(advanceCutoffs(start, [t], 'both')).toEqual(start);
  });

  it('reads the pre-C2 single {cutoff} as BOTH marks — it was written by kind-blind eviction', () => {
    expect(readCutoffs({ cutoff: '2026-01-05T08:30:00.000Z' })).toEqual({
      text: '2026-01-05T08:30:00.000Z',
      images: '2026-01-05T08:30:00.000Z',
    });
    expect(readCutoffs({ text: 'a', images: null })).toEqual({ text: 'a', images: null });
    expect(readCutoffs(null)).toEqual(NO_CUTOFFS);
    expect(readCutoffs({ cutoff: '' })).toEqual(NO_CUTOFFS);
  });
});

// ── The store's three triggers, through the ONE deleter ─────────────────────

class RecordingTransport implements TimelineTransport {
  pictures: string[] = [];
  dropRowImages(ids: readonly string[]): void {
    this.pictures.push(...ids);
  }
  reInjectLocally = vi.fn(async () => null);
  rowImage = vi.fn(async () => null);
}

function kvStore(): ReportingKvStore {
  const m = new Map<string, string>();
  return {
    get: (k: string) => m.get(k) ?? null,
    set: (k: string, v: string) => {
      m.set(k, v);
      return true;
    },
  };
}

describe('TimelineStore — 三个触发器一条删除路径 (16 册 §6.2-1)', () => {
  let transport: RecordingTransport;
  let kv: ReportingKvStore;

  beforeEach(() => {
    transport = new RecordingTransport();
    kv = kvStore();
  });

  function seeded(rows: TimelineRow[]): TimelineStore {
    kv.set('flowmic.history.cache', JSON.stringify(rows));
    return new TimelineStore(transport, kv, () => NOW);
  }

  it('🔴 G-21 — 用户删单行 now drops the picture file, exactly like 自动裁剪 does', () => {
    const pic = row({ id: 'p1', entry_type: 'image', full_image: true });
    const store = seeded([pic]);
    expect(transport.pictures).toEqual([]); // positive control: nothing dropped at boot
    expect(store.remove('p1', 'lan')).toBe(true);
    expect(transport.pictures).toEqual(['p1']);
  });

  it('the two triggers have the SAME side effects on the same row', () => {
    const pic = row({ id: 'same', entry_type: 'image', full_image: true });
    const byDelete = new RecordingTransport();
    const kvA = kvStore();
    kvA.set('flowmic.history.cache', JSON.stringify([pic]));
    new TimelineStore(byDelete, kvA, () => NOW).remove('same', 'lan');

    const byClear = new RecordingTransport();
    const kvB = kvStore();
    kvB.set('flowmic.history.cache', JSON.stringify([pic]));
    new TimelineStore(byClear, kvB, () => NOW).clear('images', 'all');

    expect(byClear.pictures).toEqual(byDelete.pictures);
    expect(JSON.parse(kvB.get('flowmic.history.cache') ?? 'x')).toEqual(
      JSON.parse(kvA.get('flowmic.history.cache') ?? 'y'),
    );
  });

  it('清空 removes exactly what previewClear promised — one selector, two callers', () => {
    const rows = [
      row({ id: 'a', created_at: '2026-01-01T00:00:00.000Z' }),
      row({ id: 'b', created_at: '2026-08-01T00:00:00.000Z' }),
      row({ id: 'c', created_at: '2026-01-02T00:00:00.000Z', entry_type: 'image', full_image: true }),
    ];
    const store = seeded(rows);
    const preview = store.previewClear('text', 'month');
    expect(preview.map((r) => r.id)).toEqual(['a']);
    expect(store.clear('text', 'month').removed).toBe(preview.length);
    expect(store.allRows().map((r) => r.id).sort()).toEqual(['b', 'c']);
    // The picture was NOT touched by a text clear — the two-way choice (text/images)
    // can be exercised independently.
    expect(transport.pictures).toEqual([]);
  });

  it('a clear that matches nothing removes nothing, persists nothing, moves no mark', () => {
    const store = seeded([row({ id: 'recent', created_at: '2026-08-02T00:00:00.000Z' })]);
    const before = kv.get('flowmic.history.cache');
    expect(store.clear('text', 'year').removed).toBe(0);
    expect(store.retention.cutoffs).toEqual(NO_CUTOFFS);
    expect(kv.get('flowmic.history.cache')).toBe(before);
  });

  it('🔴 the cleared range survives a restart —「清掉了」must not become「本来就没有」', () => {
    const store = seeded([
      row({ id: 'gone', created_at: '2026-01-01T00:00:00.000Z' }),
      row({ id: 'kept', created_at: '2026-08-01T00:00:00.000Z' }),
    ]);
    store.clear('both', 'month');
    const mark = store.retention.cutoff;
    expect(mark).toBe('2026-01-01T00:00:00.000Z');

    const reborn = new TimelineStore(new RecordingTransport(), kv, () => NOW);
    expect(reborn.retention.cutoff).toBe(mark);
    expect(reborn.retention.kept).toBe(1);
  });

  it('clearing only pictures leaves the combined sentence unsaid, and says the picture one', () => {
    const store = seeded([
      row({ id: 't', created_at: '2026-01-01T00:00:00.000Z' }),
      row({ id: 'p', created_at: '2026-01-02T00:00:00.000Z', entry_type: 'image', full_image: true }),
    ]);
    store.clear('images', 'month');
    expect(store.retention.cutoff).toBeNull(); // never claim the transcripts went too
    expect(store.retention.cutoffs.images).toBe('2026-01-02T00:00:00.000Z');
    expect(store.retention.cutoffs.text).toBeNull();
  });

  it('自动裁剪 still drops pictures — the trigger moved, the behaviour did not (RV-93)', () => {
    const many: TimelineRow[] = [];
    for (let i = 0; i <= MAX_ROWS; i++) {
      const at = new Date(Date.parse('2026-01-01T00:00:00.000Z') + i * 60_000).toISOString();
      many.push(row({ id: `r${i}`, created_at: at, entry_type: 'image', full_image: true }));
    }
    seeded(many);
    expect(transport.pictures).toEqual(['r0']);
  });
});

File: apps/mobile/test/timeline_purge_test.dart
// 16 册 §6.2 (clear) + 15 册 G-21 (single-row delete does not delete bytes) —— phone side.
//
// The most important group is 「one delete path」: single-row delete and clear
// must have **exactly the same side effects** on the same row, because before
// this card they did not — `TimelineStore.delete` only deleted the row in the
// table; the picture file and the vault's unknown fields were left alone
// (`discard(` had zero production callers).

import 'dart:typed_data';

import 'package:flowmic/src/portable/unknown_field_vault.dart';
import 'package:flowmic/src/portable/fpr_mobile.dart';
import 'package:flowmic/src/timeline/timeline_entry.dart';
import 'package:flowmic/src/timeline/timeline_persistence.dart';
import 'package:flowmic/src/timeline/timeline_purge.dart';
import 'package:flowmic/src/timeline/timeline_reaper.dart';
import 'package:flowmic/src/timeline/timeline_store.dart';
import 'package:flowmic/src/session/outbox_blob_store.dart';
import 'package:flutter_test/flutter_test.dart';

import 'support/di.dart';
import 'support/portable_rows.dart';

final DateTime _now = DateTime.utc(2026, 8, 2, 12);

TimelineEntry _row({
  required String id,
  DateTime? at,
  bool image = false,
}) => testRow(
  id: 'loc_d_$id',
  clientId: id,
  text: '一句话',
  createdAt: at ?? DateTime.utc(2026, 8, 1),
  entryType: image ? TimelineEntry.kImage : TimelineEntry.kTranscript,
);

void main() {
  group('planClear —— two kinds + time windows (owner 2026-08-01 §4-4)', () {
    final List<TimelineEntry> rows = <TimelineEntry>[
      _row(id: 'old-text', at: DateTime.utc(2026, 1, 1)),
      _row(id: 'new-text', at: DateTime.utc(2026, 8, 2)),
      _row(id: 'old-pic', at: DateTime.utc(2026, 1, 1), image: true),
      _row(id: 'new-pic', at: DateTime.utc(2026, 8, 2), image: true),
    ];

    test('「clear pictures · older than a month」 takes only the old picture; the other three stay put', () {
      final List<TimelineEntry> doomed = planClear(
        rows,
        ClearKind.images,
        horizonOf(ClearWindow.month, _now),
      );
      expect(doomed.map((TimelineEntry e) => e.clientId), <String>['old-pic']);
    });

    test('「clear text · older than a month」 is its mirror', () {
      final List<TimelineEntry> doomed = planClear(
        rows,
        ClearKind.text,
        horizonOf(ClearWindow.month, _now),
      );
      expect(doomed.map((TimelineEntry e) => e.clientId), <String>['old-text']);
    });

    test('「all」 has no lower bound; even the newest are taken', () {
      expect(planClear(rows, ClearKind.both, horizonOf(ClearWindow.all, _now)), hasLength(4));
      expect(horizonOf(ClearWindow.all, _now), isNull);
    });

    test('already-soft-deleted rows no longer count toward 「will delete N」', () {
      final TimelineEntry ghost = _row(id: 'ghost', at: DateTime.utc(2026, 1, 1)).copyWith(deleted: true);
      expect(planClear(<TimelineEntry>[ghost], ClearKind.both, horizonOf(ClearWindow.year, _now)), isEmpty);
    });

    test('the five windows get earlier in order, no shuffling', () {
      final List<DateTime> marks = <ClearWindow>[
        ClearWindow.week,
        ClearWindow.month,
        ClearWindow.quarter,
        ClearWindow.halfYear,
        ClearWindow.year,
      ].map((ClearWindow w) => horizonOf(w, _now)!).toList();
      for (int i = 1; i < marks.length; i++) {
        expect(marks[i].isBefore(marks[i - 1]), isTrue, reason: '$i');
      }
    });
  });

  group('cutoffs —— cleared ≠ never existed, and must not over-claim (16 册 §6.2-3)', () {
    final TimelineEntry t = _row(id: 't', at: DateTime.utc(2026, 3, 1));
    final TimelineEntry p = _row(id: 'p', at: DateTime.utc(2026, 4, 1), image: true);

    test('single-row delete advances no mark —— 「I deleted this one」 ≠ 「everything older than it is gone」', () {
      expect(advanceCutoffs(Cutoffs.none, <TimelineEntry>[t], null).isEmpty, isTrue);
    });

    test('🔴 when only pictures are cleared, the UI must not claim the text is gone too', () {
      final Cutoffs after = advanceCutoffs(Cutoffs.none, <TimelineEntry>[p], ClearKind.images);
      expect(after.images, p.createdAt);
      expect(after.text, isNull);
      // A single scalar cutoff would say here 「records older than April have been
      // cleared」, while every March text row is still on screen — that is exactly
      // why the marks are split by type.
      expect(after.combined, isNull);
    });

    test('a type-agnostic delete counts both marks (even if this pass only scooped text)', () {
      final Cutoffs after = advanceCutoffs(Cutoffs.none, <TimelineEntry>[t], ClearKind.both);
      expect(after.text, t.createdAt);
      expect(after.images, t.createdAt);
      expect(after.combined, t.createdAt);
    });

    test('marks only move forward', () {
      const Cutoffs start = Cutoffs(text: null, images: null);
      final Cutoffs a = advanceCutoffs(start, <TimelineEntry>[p], ClearKind.both);
      final Cutoffs b = advanceCutoffs(a, <TimelineEntry>[t], ClearKind.both);
      expect(b.text, p.createdAt); // March is earlier than April; the mark does not retreat
    });

    test('persist round-trip: still answers after a restart (JSON shape)', () {
      final Cutoffs before = advanceCutoffs(Cutoffs.none, <TimelineEntry>[t, p], ClearKind.both);
      final Cutoffs after = Cutoffs.fromJson(before.toJson());
      expect(after.text, before.text);
      expect(after.images, before.images);
      expect(Cutoffs.fromJson(null).isEmpty, isTrue);
    });
  });

  group('TimelineReaper —— deleting a row = deleting every byte of that row (16 册 §6.2-2)', () {
    late InMemoryOutboxBlobStore images;
    late InMemoryUnknownFieldVault vault;
    late InMemoryTimelinePersistence persistence;
    late TimelineReaper reaper;

    setUp(() async {
      images = InMemoryOutboxBlobStore();
      vault = InMemoryUnknownFieldVault();
      persistence = InMemoryTimelinePersistence();
      reaper = TimelineReaper(
        persistence: persistence,
        images: images,
        vault: vault,
        cutoffs: InMemoryCutoffStore(),
      );
    });

    test('🔴 the row is gone ∧ the picture file is gone ∧ the vault entry is gone —— all three must be asserted', () async {
      final TimelineEntry pic = _row(id: 'p1', image: true);
      await persistence.upsert(pic);
      await images.put(
        requestId: 'p1',
        bytes: Uint8List.fromList(List<int>.filled(500, 7)),
        extension: 'png',
      );
      await vault.merge(<String, FprCarriedFields>{
        pic.id: const FprCarriedFields(
          top: <String, Object?>{'future_field': 1},
          ext: <String, Object?>{},
        ),
      });
      // Positive control: all three things are there before we act.
      expect(await persistence.loadAll(), hasLength(1));
      expect(await images.pathFor('p1'), isNotNull);
      expect((await vault.readAll()).containsKey(pic.id), isTrue);

      final ReapResult out = await reaper.reap(<TimelineEntry>[pic]);

      expect(out.rows, 1);
      expect(out.pictures, 1);
      expect(out.bytesFreed, 500); // measured, not pictures × an average (§6.2-5)
      expect(await persistence.loadAll(), isEmpty);
      expect(await images.pathFor('p1'), isNull);
      expect((await vault.readAll()).containsKey(pic.id), isFalse);
    });

    test('an image row with no picture file does not promise to free bytes that do not exist (§9b-6)', () async {
      final TimelineEntry pic = _row(id: 'p2', image: true);
      await persistence.upsert(pic);
      final ReapResult out = await reaper.reap(<TimelineEntry>[pic]);
      expect(out.rows, 1);
      expect(out.pictures, 0);
      expect(out.bytesFreed, 0);
    });

    test('single-row delete does not write a cutoff; only clear does', () async {
      final TimelineEntry e = _row(id: 't1');
      await persistence.upsert(e);
      await reaper.reap(<TimelineEntry>[e]);
      expect(reaper.cutoffs.isEmpty, isTrue);

      final TimelineEntry e2 = _row(id: 't2');
      await persistence.upsert(e2);
      await reaper.reap(<TimelineEntry>[e2], advance: ClearKind.both);
      expect(reaper.cutoffs.combined, e2.createdAt);
    });
  });

  group('TimelineStore —— two triggers, one delete path (16 册 §6.2-1)', () {
    late InMemoryOutboxBlobStore images;
    late InMemoryUnknownFieldVault vault;
    late InMemoryTimelinePersistence persistence;
    late TimelineStore store;

    Future<TimelineStore> seeded(List<TimelineEntry> rows) async {
      images = InMemoryOutboxBlobStore();
      vault = InMemoryUnknownFieldVault();
      persistence = InMemoryTimelinePersistence();
      for (final TimelineEntry e in rows) {
        await persistence.upsert(e);
        if (e.isImage) {
          await images.put(
            requestId: e.clientId,
            bytes: Uint8List.fromList(List<int>.filled(300, 1)),
            extension: 'png',
          );
        }
      }
      final TimelineStore s = TimelineStore(
        persistence: persistence,
        reaper: newTestReaper(
          persistence: persistence,
          images: images,
          vault: vault,
        ),
      );
      await s.load();
      return s;
    }

    test('🔴 G-21 —— user deleting a single row now also deletes the picture file', () async {
      final TimelineEntry pic = _row(id: 'p1', image: true);
      store = await seeded(<TimelineEntry>[pic]);
      addTearDown(store.dispose);
      expect(await images.pathFor('p1'), isNotNull); // positive control

      store.delete(pic.id);
      // delete is fire-and-forget (same posture as this class's other persist);
      // wait for it to be scheduled.
      await Future<void>.delayed(Duration.zero);
      await Future<void>.delayed(Duration.zero);

      expect(store.entries, isEmpty);
      expect(await images.pathFor('p1'), isNull);
    });

    test('the two triggers have exactly the same side effects on the same row', () async {
      final TimelineEntry pic = _row(id: 'same', image: true);

      final TimelineStore byDelete = await seeded(<TimelineEntry>[pic]);
      byDelete.delete(pic.id);
      await Future<void>.delayed(Duration.zero);
      await Future<void>.delayed(Duration.zero);
      final bool fileGoneA = await images.pathFor('same') == null;
      final List<TimelineEntry> leftA = await persistence.loadAll();
      byDelete.dispose();

      final TimelineStore byClear = await seeded(<TimelineEntry>[pic]);
      await byClear.clear(ClearKind.images, ClearWindow.all, now: _now);
      final bool fileGoneB = await images.pathFor('same') == null;
      final List<TimelineEntry> leftB = await persistence.loadAll();
      byClear.dispose();

      expect(fileGoneB, fileGoneA);
      expect(fileGoneA, isTrue);
      expect(leftB.length, leftA.length);
      expect(leftA, isEmpty);
    });

    // D7 (0.3.0) — this test originally asked `store.previewClear(...)`, and that
    // method had **zero production callers** and only looked at rows already
    // paged into memory; it was deleted with D7 (reason written at the original
    // site in `timeline_store.dart`). The **intent** of the assertion did not
    // change; what changed is who is asked: now it asks the estimate chain the
    // clear panel actually uses (`AssetInventory.readAllRows()` →
    // [TimelineStore.readAllRowsForInventory] → [planClear]), i.e. the same
    // whole-store source the reaper now picks rows from. The half that crosses
    // a pagination boundary lives in timeline_clear_boundary_test.dart
    // (>60 rows; the old implementation went red there).
    test('clear deletes exactly the rows the 「estimate chain」 promised (same predicate, two callers)', () async {
      final List<TimelineEntry> rows = <TimelineEntry>[
        _row(id: 'a', at: DateTime.utc(2026, 1, 1)),
        _row(id: 'b', at: DateTime.utc(2026, 8, 1)),
        _row(id: 'c', at: DateTime.utc(2026, 1, 2), image: true),
      ];
      store = await seeded(rows);
      addTearDown(store.dispose);

      final List<TimelineEntry> preview = planClear(
        await store.readAllRowsForInventory(),
        ClearKind.text,
        horizonOf(ClearWindow.month, _now),
      );
      expect(preview.map((TimelineEntry e) => e.clientId), <String>['a']);

      final ReapResult out = await store.clear(ClearKind.text, ClearWindow.month, now: _now);
      expect(out.rows, preview.length);
      expect(
        store.entries.map((TimelineEntry e) => e.clientId).toList()..sort(),
        <String>['b', 'c'],
      );
      // Clearing text did not touch pictures —— the two kinds can run separately.
      expect(await images.pathFor('c'), isNotNull);
    });

    test('when the range holds nothing, no row is deleted and the marks do not move', () async {
      store = await seeded(<TimelineEntry>[_row(id: 'recent', at: DateTime.utc(2026, 8, 2))]);
      addTearDown(store.dispose);
      final ReapResult out = await store.clear(ClearKind.text, ClearWindow.year, now: _now);
      expect(out.rows, 0);
      expect(store.cutoffs.isEmpty, isTrue);
      expect(store.entries, hasLength(1));
    });

    test('🔴 after a clear, the sentence 「what range was cleared」 still holds after the store is reconstructed', () async {
      final InMemoryCutoffStore disk = InMemoryCutoffStore();
      images = InMemoryOutboxBlobStore();
      vault = InMemoryUnknownFieldVault();
      persistence = InMemoryTimelinePersistence();
      final TimelineEntry gone = _row(id: 'gone', at: DateTime.utc(2026, 1, 1));
      final TimelineEntry kept = _row(id: 'kept', at: DateTime.utc(2026, 8, 1));
      await persistence.upsert(gone);
      await persistence.upsert(kept);

      TimelineStore build() => TimelineStore(
        persistence: persistence,
        reaper: newTestReaper(persistence: persistence, images: images, vault: vault, cutoffs: disk),
      );

      final TimelineStore first = build();
      await first.load();
      await first.clear(ClearKind.both, ClearWindow.month, now: _now);
      final DateTime? mark = first.cutoffs.combined;
      expect(mark, gone.createdAt);
      first.dispose();

      // Reopen once (the same 「disk」): that sentence must still be there.
      final TimelineStore reborn = build();
      await reborn.load();
      addTearDown(reborn.dispose);
      expect(reborn.cutoffs.combined, mark);
      expect(reborn.entries, hasLength(1));
    });
  });
}
